### ISSUE TITLE
Build error with CUDA 10.2 Jetson Nano

### ISSUE THREAD

**Issue by alexismailov2** [OpenNMT/CTranslate2#1378]
Please help how can I use CTranslate2 with CUDA 10.2?
The problem is in the Nvidia Jetson Nano does not work with CUDA 11 and 12(installation of cuda 11 or cuda 12 help to build but it does not work due to driver issue)…
Cuda 10.2 will work well but looks like CTranslate2 needs at least CUDA 11…
Provide solution how to fix CTranslate2 to build it for CUDA 10.2…

**Comment by guillaumekln** [comment]
CUDA 10 support was dropped 1 year ago.

What build error do you get? I think the main incompatibility with CUDA 10 is the usage of the C++17 standard.

https://github.com/OpenNMT/CTranslate2/blob/fd93a1f5548d0bcf60ab756cca2b37c109d54592/CMakeLists.txt#L426

You can try using `c++14` in this line but it may require to some changes in the code also.

**Comment by alexismailov2** [comment]
During compiling I have gotten errors chained with missed headers If you need to understand what actually I can add here this info. But I think from your side will be better to create a CI/CD runner for NVIDIA Jetson Nano and you will see that problems.
Could you provide a fork to keep support of building for NVIDIA Jetson Nano, because that pcb is enough power compared to another pcb in that price range(54$ for NVIDIA Jetson Nano 2Gb version).
With installed CUDA 12.2 on Jetson Nano it build but not run due to some problems with internal driver which looks like was not updated by Nvidia for Jetson Nano.

**Comment by juansebashr** [comment]
Sorry @alexismailov2, have you encountered a solution? I'm facing the same problem

**Comment by alexismailov2** [comment]
@juansebashr Unfortunately still no, currently using Nvidia Jetson Orin Nano with cuda 11 support(expensive board but enough efficient). 
In addition I plan to introduce support of rk3588 NPU(Orange Pi 5, 6 TOPS) hope it will not be very difficult but could not say now anything…

**Comment by juansebashr** [comment]
@alexismailov2 Sadly I can't change the board :( haha, I will give it a try and write a comment if I accomplish anything

**Comment by alexismailov2** [comment]
I tried to build with cuda 10.2 but it produced an error with missed includes, may be localization of this dependency can be reduced…

**Comment by MuhammadShifa** [comment]
Hello @alexismailov2 ,
I am also try to build with cuda 11.4 on Jeson Orin Nano but facing issue doing cmake ..
CMake Error at CMakeLists.txt:273 (message):
  Intel OpenMP runtime libiomp5 not found

Please help me if you have solved this issue.

**Comment by alexismailov2** [comment]
You can turn it off with cmake option, I have build without problem it…

**Comment by alexismailov2** [comment]
-DWITH_MKL=OF

**Comment by MuhammadShifa** [comment]
Thank you @alexismailov2 ,
Yes I need to add this option -DWITH_MKL=OFF, but I don't know where to add this option. Although in main directory of ctranslae2 CMakeLists.txt I have turn off:  option(WITH_MKL "Compile with Intel MKL backend" OFF) but still it showing me this error.

**Comment by alexismailov2** [comment]
cmake -Bbuild_folder -DWITH_MKL=OFF
cmake —build build_folder

**Comment by MuhammadShifa** [comment]
I tried multiple command making enabled with cuda but still facing issue. Cranslate2 is the repository clone while I have `mkdir build && cd build` command and then `cmake` facing issue
![Screenshot from 2023-12-27 15-23-11](https://github.com/OpenNMT/CTranslate2/assets/36985668/b866c362-0924-46b6-be93-df69209df1c8)

**Comment by alexismailov2** [comment]
Add option -DOPENMP_RUNTIME=NONE

**Comment by MuhammadShifa** [comment]
Thank you so much @alexismailov2 .
I have run the above command successfully. After this I am running the `sudo make install` command and it also working fine now but the system is in progress in installation. Thanks a lot for your precious time.

**Comment by alexismailov2** [comment]
You can even create wheel python package, for installation

**Comment by alexismailov2** [comment]
And be ready to understand that nvidia orin nano give good comparison only on big models and medium, on tiny unfortunately OrangePi 5 will be faster…

**Comment by MuhammadShifa** [comment]
I appreciate your help. I would be happy if you share me a resource to create a wheel.
Installation done 100% but again I am facing issue:
`e.py", line 10, in <module>
    import ctranslate2
ModuleNotFoundError: No module named 'ctranslate2'`

I also want to run FasterWhisper medium model. It's working fine on my window based system but in Jetson Orin Nano I am facing this issue.

**Comment by alexismailov2** [comment]
Give me a second I will try to remember

**Comment by MuhammadShifa** [comment]
Yes I also follow these steps from [here](https://opennmt.net/CTranslate2/installation.html#build-options). I will update you soon.

**Comment by MuhammadShifa** [comment]
I followed your every steps and installed successfully. But I have face issue of cuda:
`ValueError: This CTranslate2 package was not compiled with CUDA support`

I think I need to pass -DWITH_CUDA=ON then I think it will work. If you have any suggestion, Please tell me I think that's the reason it showed me this error.

**Comment by alexismailov2** [comment]
I need to open my nvidia jetson nano and remember how to do that… I remember that I habe spend on it some time…

**Comment by MuhammadShifa** [comment]
Thanks @alexismailov2 , 
I am hopeful the above last command work. I will update you. I am already working you don't need to open jetson. Thanks a lot.

**Comment by alexismailov2** [comment]
If you work in some company which might need my expertise you can suggest me as a potential contractor!

**Comment by PhilipAmadasun** [comment]
I have made a docker image with successfully built ctranslate2 for jetpack 6. The base image is from the jetson-containers pytorch github repo.  I had to downgrade the gcc compiler to version 10.

**Comment by ArmanJR** [comment]
A slightly improved version of the above steps, with a fix on `TypeError: canonicalize_version() got an unexpected keyword argument 'strip_trailing_zero`:

## Install CUDA enabled CTranslate2

```bash
git clone --recursive https://github.com/OpenNMT/CTranslate2.git
cd CTranslate2
cmake -Bbuild_folder -DWITH_MKL=OFF -DOPENMP_RUNTIME=NONE -DWITH_CUDA=ON -DWITH_CUDNN=ON
cmake —build build_folder
cd build_folder
sudo make install
cd ../python
python3 -m venv venv
source venv/bin/activate
pip install --upgrade pip setuptools wheel packaging
pip install -r install_requirements.txt
python3 setup.py bdist_wheel
pip install dist/*.whl
echo "$PWD/"*
```

The last command will print the full path of the CTranslate2 wheel, which you can later use to install CTranslate2 on any virtual env via:

```bash
pip install [path to .whl file]
```

**Comment by End46** [comment]
> CUDA 10 support was dropped 1 year ago.
> 
> What build error do you get? I think the main incompatibility with CUDA 10 is the usage of the C++17 standard.
> 
> [CTranslate2/CMakeLists.txt](https://github.com/OpenNMT/CTranslate2/blob/fd93a1f5548d0bcf60ab756cca2b37c109d54592/CMakeLists.txt#L426)
> 
> Line 426 in [fd93a1f](/OpenNMT/CTranslate2/commit/fd93a1f5548d0bcf60ab756cca2b37c109d54592)
> 
>  list(APPEND CUDA_NVCC_FLAGS "-std=c++17") 
> You can try using `c++14` in this line but it may require to some changes in the code also.

Hi, maybe you could help me. I'm experiencing the same issue, am trying to build version 4.6.0 on Cuda 10.2 to use faster-whisper on a Jetson Nano Developer Kit. I changed that line in the cmake list and it continues with the build, but I know it will throw up some errors in the code. The problem is, I don't know where to start looking for these problems. Could you give me some guidance?

**Comment by pipi32167** [comment]
Thanks for @MuhammadShifa, I have a working Dockerfile in DGX Spark:

```

FROM nvidia/cuda:12.3.2-cudnn9-devel-ubuntu22.04 AS builder

WORKDIR /root

# Install essential build dependencies
RUN apt-get update && apt-get install -y \
  python3-pip \
  git \
  cmake \
  build-essential \
  libopenblas-dev \
  && rm -rf /var/lib/apt/lists/*

# Clone CTranslate2 repository with submodules
RUN git clone --recursive https://github.com/OpenNMT/CTranslate2.git

# Build and install CTranslate2 with CUDA and cuDNN support
WORKDIR /root/CTranslate2
RUN cmake -B build_folder \
  -DWITH_MKL=OFF \
  -DOPENMP_RUNTIME=NONE \
  -DWITH_CUDA=ON \
  -DWITH_CUDNN=ON \
  && cmake --build build_folder \
  && cd build_folder \
  && make install \
  && ldconfig

COPY infer.py jfk.flac /root

RUN pip3 install faster-whisper

# Build and install Python bindings (wheel)
WORKDIR /root/CTranslate2/python
RUN pip3 install -r install_requirements.txt \
  && python3 setup.py bdist_wheel \
  && pip3 install --force-reinstall dist/*.whl

WORKDIR /root

# Default command
CMD ["python3", "infer.py"]
```
